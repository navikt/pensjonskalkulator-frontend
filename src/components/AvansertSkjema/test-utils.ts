export function getPreviousMonth(): string {
  const today = new Date()
  const currentMonth = parseInt(
    today.toLocaleDateString('en-CA').slice(5, 7),
    10
  ) // Extracts the month from the date string
  const previousMonth = currentMonth - 1
  return previousMonth.toString()
}
