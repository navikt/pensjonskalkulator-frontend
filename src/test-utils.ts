import renderer from 'react-test-renderer'

import { render } from '@testing-library/react'

export function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
}

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => children,
    ...options,
  })

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
