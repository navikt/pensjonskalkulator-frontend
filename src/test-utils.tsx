import React, { PropsWithChildren } from 'react'
import { IntlProvider } from 'react-intl'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'

import { PreloadedState, createListenerMiddleware } from '@reduxjs/toolkit'
import { render, RenderOptions } from '@testing-library/react'

import { getTranslation_test } from '@/utils/__tests__/test-translations'

import { createUttaksalderListener } from './state/listeners/uttaksalderListener'
import {
  setupStore,
  RootState,
  AppStore,
  AppStartListening,
} from './state/store'
import { getTranslation_nb } from './translations/nb'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
  hasRouter?: boolean
}

export const swallowErrors = (testFn: () => void) => {
  const error = console.error
  console.error = () => {}
  testFn()
  console.error = error
}

export const swallowErrorsAsync = async (testFn: () => Promise<void>) => {
  const cache = console.error
  console.error = () => {}
  await testFn()
  console.error = cache
}

function generateMockedTranslations() {
  const nbTranslations: Record<string, string> = getTranslation_nb()
  const testTranslations: Record<string, string> = getTranslation_test()
  const translationsInput = { ...nbTranslations, ...testTranslations }
  const translations: Record<string, string> = {}

  for (const key in translationsInput) {
    if (
      /<(?=.*? .*?\/ ?>|br|hr|input|!--|wbr)[a-z]+.*?>|<([a-z]+).*?<\/\1>|{[^}]+}/i.test(
        translationsInput[key]
      )
    ) {
      // for html or chunks keys: results in 'my_key' : 'my_key with some <html> or {chunk}'
      translations[key] = translationsInput[key]
    } else {
      // results in 'my_key' : 'my_key'
      translations[key] = key
    }
  }
  return translations
}

// Return an object with the store and all of RTL's query functions
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState, true),
    hasRouter = true,
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<unknown>): JSX.Element {
    return (
      <Provider store={store}>
        <IntlProvider locale="nb" messages={generateMockedTranslations()}>
          {hasRouter ? <MemoryRouter>{children}</MemoryRouter> : children}
        </IntlProvider>
      </Provider>
    )
  }

  const listenerMiddleware = createListenerMiddleware()
  createUttaksalderListener(
    listenerMiddleware.startListening as AppStartListening
  )

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'

export { renderWithProviders as render }
