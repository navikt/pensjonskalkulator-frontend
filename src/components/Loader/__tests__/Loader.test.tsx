import { describe, it } from 'vitest'

import { Loader } from '..'
import { render, screen } from '@/test-utils'

describe('Loader', () => {
  it('rendrer med riktig className og props by default', () => {
    const { asFragment } = render(
      <Loader title="fake title" data-testid="loader" />
    )

    expect(screen.getByTestId('loader')).toBeVisible()
    expect(asFragment()).toMatchSnapshot()
  })

  it('rendrer med riktig className nÃ¥r isCentered er false', () => {
    const { asFragment } = render(
      <Loader title="fake title" data-testid="loader" isCentered={false} />
    )

    expect(screen.getByTestId('loader')).toBeVisible()
    expect(asFragment()).toMatchSnapshot()
  })
})
