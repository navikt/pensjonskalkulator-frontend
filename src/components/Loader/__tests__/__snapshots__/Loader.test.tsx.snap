// Vitest Snapshot v1

exports[`Loader > rendrer med riktig className n√•r isCentered er false 1`] = `
<DocumentFragment>
  <svg
    aria-labelledby="loader-r1"
    class="navds-loader navds-loader--medium navds-loader--neutral"
    data-testid="loader"
    focusable="false"
    preserveAspectRatio="xMidYMid"
    viewBox="0 0 50 50"
  >
    <title
      id="loader-r1"
    >
      fake title
    </title>
    <circle
      class="navds-loader__background"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      xmlns="http://www.w3.org/2000/svg"
    />
    <circle
      class="navds-loader__foreground"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      stroke-dasharray="50 155"
    />
  </svg>
</DocumentFragment>
`;

exports[`Loader > rendrer med riktig className og props by default 1`] = `
<DocumentFragment>
  <svg
    aria-labelledby="loader-r0"
    class="navds-loader _loader__isCentered_d51639 navds-loader--medium navds-loader--neutral"
    data-testid="loader"
    focusable="false"
    preserveAspectRatio="xMidYMid"
    viewBox="0 0 50 50"
  >
    <title
      id="loader-r0"
    >
      fake title
    </title>
    <circle
      class="navds-loader__background"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      xmlns="http://www.w3.org/2000/svg"
    />
    <circle
      class="navds-loader__foreground"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      stroke-dasharray="50 155"
    />
  </svg>
</DocumentFragment>
`;
